### PR TITLE
Revise Stores API, with tests and OAS

### PR DESCRIPTION
**Pull request by mshappe**
This does not yet cover the product sub-API, just Stores crud. Now that I've found a good path, patterns, and rhythm, Product should be easy to revise, but I ran out of oomph.

### CODE REVIEW

**Comment by mshappe** [comment]
This also contains a trial balloon for a pattern we can apply for name conversions in and out.

File: api/src/Controller/Stores/BaseStore.php [hunk 1]
```diff
@@ -13,24 +74,75 @@
...
     protected $id = 0;
 
+    /* This the various mapping functions here are intended as a
+     * prototype for how we could handle name conversion. It's
+     * the beginning of an idea, not the end of it.
+     */
+    protected static $columnsToParams = [
+        'StoreID' => 'id',
+        'StoreSlug' => 'store_slug',
+        'Name' => 'name',
```
[the remark this answers is not in the document]
Comment (mshappe) [reply api/src/Controller/Stores/BaseStore.php:90]: I really, really, really hate this rule. Like, really hate.

**Comment by AricStewart** [comment]
I really like this a lot... just a few overarching comments I would wonder about.  If you say "do that yourself next" that is valid.

TL;DR:  Squash please then awesome and we can work from there.

1) Phplint HATES the splat operator, well phplint hate a lot of our code which is why i have not tried to be phplint clean. Or at least not really...    Quick web search seem to imply the splat operator is generally viewed as good and kosher so that rocks!  I had not heard of it before. Very cool!

2) Squash please?  I will admit I am old fashion and I am looking at things in the actually files and using git log and git show in my terminal and when something is "commit, fix commit, fix commit etc..." then it is REALLY hard to see the final proper code.  I would really prefer to have the fixes squashed into the original commit before the pull request is completed.

3) I am really loving the body/payload/mapping infrastructure you are building.  With `columnsToParams` and `paramsToColumns` and `insertPayloadFromParams` etc... I am already thinking about how to generalize it and be able to pull it out into the general architecture. 

4) Only point is there are things like `filterBodyParams`, which is always used on a body?  Seems like the new code path is that it is being used by `insertPayloadFromParams` which is used on more than just `body`   Maybe we rename?  just `filterParams` sounds good.  or `filterArray`  Again I am thinking generalizing...

**Review by AricStewart — APPROVED** [review]

**Comment by mshappe** [comment]
> 1. Phplint HATES the splat operator, well phplint hate a lot of our code which is why i have not tried to be phplint clean. Or at least not really...    Quick web search seem to imply the splat operator is generally viewed as good and kosher so that rocks!  I had not heard of it before. Very cool!

At some point, we may need to revisit phplint/phpcs rulesets to accommodate our actual preferred practices. They do not have to be prescriptive (let alone proscriptive)--every linter I know has a way to tell it, "No, we actually like that one."

The splat operator is also relatively new, and I'm guessing (but really don't know) that it was controversial when introduced, given how old PHP is and the likelihood of conservatism in its ranks. But it's EXTREMELY handy in a situation like this one, where we have a function that doesn't take an array, but a variadic list of parameters; but what we have to PASS it is an array. Every modern language has one, now. JavaScript also uses `...`; Ruby uses `*` (which may be where "splat" came from for the function).

> 3. I am really loving the body/payload/mapping infrastructure you are building.  With `columnsToParams` and `paramsToColumns` and `insertPayloadFromParams` etc... I am already thinking about how to generalize it and be able to pull it out into the general architecture.

Yeah, I can see either pulling it out, or just making it a conventional pattern if it turns out there's not much actual logic to hoist safely, but I actually think both `insertPayload` and `selectMapping` could be hoisted, and wrote them as generically as possible to support the idea.

> 4. Only point is there are things like `filterBodyParams`, which is always used on a body?  Seems like the new code path is that it is being used by `insertPayloadFromParams` which is used on more than just `body`   Maybe we rename?  just `filterParams` sounds good.  or `filterArray`  Again I am thinking generalizing...

`filterBodyParams` was always used on a body, but that was an accident, honestly. The body in question was always an assoc-array generated from a JSON payload, which means any assoc-array will do. Name probably needs to change, because naming things is hard, but "body" was merely a question of how it was being used at the time.

**Comment by mshappe** [comment]
There's still a part of me that thinks we should use middleware for all the renaming, but I also realized that our column names are simply too all-over-the-place. So I went with the next best thing: a single explicitly written map of columns to parameters; a second map that's just the inverse; and two easy functions for using them. There may be other use cases beyond those--I definitely didn't foreclose possibilities, there--but that seemed like a reasonable least-time-to-implement compromise that allows database column names to be whatever-the-fuck they are, and our protocol parameters to fit what we WISH they were :-D

**Comment by AricStewart** [comment]
phpcs is easy to configure:  phpcs.xml    feel free to make merge requests :)

Yeah, When I signed on I repeated up and down to not let me do database design because I suck at it... Well I have been doing all the database design, based on the original design by Thomas which never had an idea where we would end up.  So the Database has naming conventions I HATE and did not bring over into the API and such.

Explicitly written maps are likely our best way, I agree with you.
